app/models/access.rb

```diff
@@ -0,0 +1,86 @@
+class Access < ApplicationRecord
+  ALLOWED_RESOURCES = %w[Organization FacilityGroup Facility].freeze
+  ACTION_TO_ROLE = {
+    view: [:manager, :viewer],
+    manage: [:manager]
+  }
+
+  belongs_to :user
+  belongs_to :resource, polymorphic: true, optional: true
+
+  enum role: {
+    viewer: "viewer",
+    manager: "manager",
+    super_admin: "super_admin"
+  }
+
+  validates :role, presence: true
+  validates :user, uniqueness: {scope: [:resource_id, :resource_type], message: "can only have one access per resource."}
+  validates :resource, presence: {unless: :super_admin?, message: "is required if not a super_admin."}
+  validates :resource, absence: {if: :super_admin?, message: "must be nil if super_admin"}
+  validates :resource_type, inclusion: {in: ALLOWED_RESOURCES, unless: :super_admin?}
+  validate :user_has_only_one_role, if: -> { user.present? }
+
+  class << self
+    def organizations(action)
+      resources_for(Organization, action)
+    end
+
+    def facility_groups(action)
+      resources_for(FacilityGroup, action)
+        .or(FacilityGroup.where(organization: organizations(action)))
+    end
+
+    def facilities(action)
+      resources_for(Facility, action)
+        .or(Facility.where(facility_group: facility_groups(action)))
+    end
+
+    def can?(action, model, record = nil)
+      if record&.is_a? ActiveRecord::Relation
+        raise ArgumentError, "record should not be an ActiveRecord::Relation."
+      end
+
+      case model
+        when :facility
+          can_access_record?(facilities(action), record)
+        when :organization
+          can_access_record?(organizations(action), record)
+        when :facility_group
+          can_access_record?(facility_groups(action), record)
+        else
+          raise ArgumentError, "Access to #{model} is unsupported."
+      end
+    end
+
+    private
+
+    def can_access_record?(resources, record)
+      return resources.exists? if super_admin?
+      return resources.find_by_id(record).present? if record
+
+      resources.exists?
+    end
+
+    def resources_for(resource_model, action)
+      return resource_model.all if super_admin?
+
+      resource_ids = where(resource_type: resource_model.to_s, role: ACTION_TO_ROLE[action]).pluck(:resource_id)
+      resource_model.where(id: resource_ids)
+    end
+
+    def super_admin?
+      super_admin.exists?
+    end
+  end
+
+  private
+
+  def user_has_only_one_role
```

Comment:
Since we have decided to only have 1 _kind of_ role per user across many accesses, we can't write a simple database constraint anymore. This would have to be an application constraint and/or some unwieldy check-constraint on pg that we should avoid for now.

It might seem a bit nonsensical to offer this abstraction in the first place, why have the schema encode roles in the Access if there's only ever going to be 1?

The only good reason to do this is future flexibility by trading-off correctness... pretty much no other reason why. It makes it unnecessarily difficult to comprehend why this validation exists in this class at all.

OTOH, the flexibility that it provides is that we can change our minds about this without needing a migration and/or write more nuanced/different cases.

Reply:
I also think we shouldn't put `role` on `Access`. A data model that _allows_ for more flexibility than our app can support means either complicated code to account for inconsistent data, or constraints we need to manage, and this isn't easily solvable with a simple unique index. Are we comfortable enough to rely on the model vaildation? Isn't this "too-flexible data model" what introduced complexity in our current permissioning system too?

For example, if I want to decide whether to show a user a "Manage" link in our nav, would I iterate through all accesses and check for manage ability? Or just the first one, assuming they're all the same?

To address the concerns we had with the "keep role on user" approach, here are my instincts, let me know if they jive with you.

**The abstraction is spread across two classes**

I think that's OK because, the rest of our app will _only_ talk to `User` anyway! With the right methods and delegation on `User`, the rest of our app shouldn't need to know about accesses at all. I think that sets us up for evolving our permissions nicely as well. The trick will be making sure that the `Access` deals only with "which resources", and the `User` deals with "which actions". 

If modeling that is a problem...
(get ready for a cheeky suggestion)
```
class Access
  delegate :role, to: :user
end
```
Let the `Access` take care of all of it, and just pull the role it needs from the user. This way our data model is correct, and our abstraction is centralized.
It means some of the methods in here will have to resort to in-memory searches instead of ActiveRecord queries, but that should be ok, since we'll be dealing with a single user's accesses at most. For example:

```ruby
resource_ids = where(resource_type: resource_model.to_s).select { |access|
  access.role.in?(ACTION_TO_ROLE[action])
}.map(&:resource_id)
```

**Naming**

I'm not a big fan of `AccessibleResource` either 😅 it's super verbose. Let's also forget about that confusing role/designation field for a second. What if users had a "role" field and many "scopes"?

Reply:
@harimohanraj89 I generally agree with this, but I think it's futile to try to achieve what we're trying to achieve with 100% success. `User` and `Access` are tightly coupled and you're going to leak information in either direction regardless of how hard you try. Unless you try and create yet [another abstraction](http://timelessrepo.com/abstraction-creep) that wraps over both.

The real cost-benefit analysis of _where access levels / roles lie_ are the following (and I still think that's the core idea we should focus on), everything else is exercise of subjective balance:

* You get clean and clear database constraints of only 1-role-per-`User` if the role is in `User`. I'd also argue that it's better, because when designing user accesses, you want strong, reliable database constraints.
* Future flexibility of changing the idea of 1 role per `User` if it's in `Access` without a migration. We've already gone back-and-forth on the 1-role-per-`User` constraint twice and introduced two more roles in the process of simplification, it isn't terribly hard to imagine this being an actual use-case and also doesn't necessarily differ much in terms of code-complexity.

To that extent, here's an attempt at moving the `access_level` (`role` was taken) to `User` but keeping some of fetching responsibilities in `Access` (as per your suggestions) – [Role in Access](https://github.com/simpledotorg/simple-server/pull/1185/files/e9aa57aa00d4fe38c776d98256b3e694cea33687) versus [Role in User](https://github.com/simpledotorg/simple-server/pull/1185/commits/fdaccce8955b444f1fe0998584a7d81204443f7b). You can compare these side-by-side and let me know what you think. Let's decide and freeze on one approach soon.

Reply:
Thanks for laying out these comparisons, it's really helpful.

My gut says "keep role in User" because
* The clean DB constraint
* Migrating isn't hard